core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1963 @@
+        match self.get_handle_type(child) {
+            Leaf => self.push_front_leaf(child, pair_ptr),
+            Internal => {
+                self.push_front_internal(
+                    child,
+                    pair_ptr,
+                    edge.expect("edge always exists for internal nodes"),
+                )
+            }
+        }
+    }
+
+    /// This removes a key/value pair from the right child and replaces it with the
+    /// key/value pair pointed to by `handle` while pushing the old key/value pair
+    /// of `handle` into the left child.
+    fn steal_right(&mut self, handle: KVHandle) {
+        let right = self.right_edge(handle);
+        let child = self.descend(right).expect("child must exist");
+        let (k, edge) = self.pop_front(child);
+
+        let node = self.get_node_mut(handle.node()).expect("node must exist");
+        let pair_ptr = node
+            .set_pair(handle.idx(), Some(k))
+            .expect("key must exist");
+
+        let left_child = self.left_child(handle).expect("left child must exist");
+        match self.get_handle_type(left_child) {
+            Leaf => self.push_leaf(left_child, pair_ptr),
+            Internal => {
+                self.push_internal(
+                    left_child,
+                    pair_ptr,
+                    edge.expect("edge always exists for internal node"),
+                )
+            }
+        }
+    }
+
+    /// Removes a pointer to a key/value pair from the beginning of this node. If this is an
+    /// internal node, also removes the edge that was to the left of that pair.
+    ///
+    /// *Note:* This method does not actually remove the storage entry! Just the pointer to it!
+    ///
+    /// Returns `(removed_storage_pointer, Option<removed_edge>)`.
+    fn pop_front(
+        &mut self,
+        handle: NodeHandle,
+    ) -> (KVStoragePointer, Option<NodeHandle>) {
+        let typ = self.get_handle_type(handle);
+        let node = self.get_node_mut(handle).expect("node must exist");
+
+        // Necessary for correctness, but this is an internal module
+        debug_assert!(node.len() > 0);
+        let old_len = node.len();
+
+        let pair_ptr = node.remove_pair_with_shift(0).expect("key must exist");
+
+        let edge = match typ {
+            Leaf => None,
+            Internal => {
+                let edge = node.remove_edge_with_shift(0).expect("edge must exist");
+
+                // Create a new `NodeHandle` to the new root
+                let new_root = NodeHandle::new(edge.node());
+                self.set_parent(new_root, None, None);
+
+                for i in 0..old_len {
+                    let h = KVHandle::new(handle, i);
+                    self.correct_parent_link(h);
+                }
+
+                Some(new_root)
+            }
+        };
+
+        let node = self.get_node_mut(handle).expect("node must exist");
+        node.set_len(node.len() - 1);
+
+        (pair_ptr, edge)
+    }
```

Comment:
Lot's of `... must exist` assertions in this function. Sometimes this indicates minor structuring smells.

Reply:
Yeah I agree. It's unfortunate because we need to fetch the edge, child and node behind a handle and each of those could potentially fail ‒ though it shouldn't if the implementation is sound.